src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -4175,11 +4175,18 @@ void EmitStackPop(params string[] args)
         /// but to ensure that significant backtracking can be stopped.  As such, we allow for up to O(n) work in the length of the input
         /// between checks, which means we emit checks anywhere backtracking is introduced, such that every check can have O(n) work
         /// associated with it.  This means checks:
-        /// - when restarting the whole match evaluation at a new index
-        /// - when backtracking backwards in a loop
-        /// - when backtracking forwards in a lazy loop
-        /// - when backtracking to the next branch of an alternation
-        /// - when processing a lookaround
+        /// - when restarting the whole match evaluation at a new index. Every match could end up doing O(n) work without a timeout
+        ///   check, and since this could then result in O(n) matches, we need a timeout check on each new position in order to
+        ///   avoid O(n^2) work without a timeout check.
+        /// - when backtracking backwards in a loop. Every backtracking step through the loop could evaluate the remainder of the
+        ///   pattern, which can lead to O(2^n) work if unchecked.
```

Comment:
what is O(2^n) ? not O(n^2) ?

Reply:
> what is O(2^n)

The worst case for backtracking.

Reply:
could you help me understand how to derive that?

Reply:
Sure. Let's say you have the pattern `^(\d+) *$`. And you try to match that against `123456789a`. That won't match, but the nested loops are going to end up backtracking through every possible combination of groupings, e.g. `(123456789)`, `(123456)(78)(9)`, etc. In the extreme, every digit will be in its own loop/group, e.g. `(1)(2)(3)(4)(5)(6)(7)(8)(9)`. Note that you can this have a group separator between every digit: with one, without one. Meaning there are two options at every position, whether to end the previous group or stay a new one. i.e. n bits, with 2^n possible combinations.